tests/php/BlockTypes/ProductCollection.php

```diff
@@ -70,9 +70,7 @@ private function initialize_merged_query( $parsed_block = array() ) {
 
 		$query = build_query_vars_from_query_block( $block, 1 );
 
-		$parsed_block['context']['query'] = $parsed_block['attrs']['query'];
-
-		return $this->block_instance->build_frontend_query( $query, $parsed_block, 1 );
```

Comment:
@imanish003 I think we're not passing the correct value here. The second argument of `ProductCollection::build_frontend_query` should a block object, not an array. Maybe we should revisit all tests that uses this `initialize_merged_query` method to check for any potential false positive.

Reply:
Good catch! Thanks, Tung 🙏 I will check it 🤝

Reply:
Let me share my findings regarding the old code that was passing an array instead of an object as the second argument to the `build_frontend_query` function. Here's the relevant code:

```tsx
private function initialize_merged_query( $parsed_block = array() ) {
    if ( empty( $parsed_block ) ) {
        $parsed_block = $this->get_base_parsed_block();
    }

    $this->block_instance->set_parsed_block( $parsed_block );

    $block          = new \stdClass();
    $block->context = $parsed_block['attrs'];

    $query = build_query_vars_from_query_block( $block, 1 );

    $parsed_block['context']['query'] = $parsed_block['attrs']['query'];

    return $this->block_instance->build_frontend_query( $query, $parsed_block, 1 );
}
```

The `build_frontend_query` function looks like this:

```tsx
tsxCopy code
public function build_frontend_query( $query, $block, $page ) {
    // If not in the context of a product collection block, return the query as is.
    $is_product_collection_block = $block->context['query']['isProductCollectionBlock'] ?? false;
    if ( ! $is_product_collection_block ) {
        return $query;
    }

    $block_context_query = $block->context['query'];
    // phpcs:ignore WordPress.DB.SlowDBQuery
    $block_context_query['tax_query'] = ! empty( $query['tax_query'] ) ? $query['tax_query'] : array();

    return $this->get_final_frontend_query( $block_context_query, $page );
}

```

At first glance, the first line of the function should throw an exception because we are accessing `$block` using the arrow operator. However, since we are using the null coalescing operator `??`, it won't throw any exception. Therefore, the value of `is_product_collection_block` becomes false, and the function returns `$query`.

### Now the next question is, how do other tests pass?

We are calling the `build_query_vars_from_query_block` function:

```tsx
$query = build_query_vars_from_query_block( $block, 1 );
```

As you can see [here](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/lib/compat/wordpress-6.2/blocks.php#L172), this function applies the filter `query_loop_block_query_vars`:

```tsx
return apply_filters( 'query_loop_block_query_vars', $query, $block, $page );
```

We use the `query_loop_block_query_vars` filter in the Product Collection block [here](https://github.com/woocommerce/woocommerce-blocks/blob/f3070cb2f12bcbe9ff46fe967eabf58764df8ce7/src/BlockTypes/ProductCollection.php#L68-L73):

```tsx
add_filter(
    'query_loop_block_query_vars',
    array( $this, 'build_frontend_query' ),
    10,
    3
);
```

As a result, our `build_frontend_query` is called with the correct values and returns the correct result, allowing all other tests to pass.

### But why did the `test_merging_filter_by_attribute_queries` test fail?

The reason is that this test sets the value of `$attributes_filter_query_args` using the following code:

```tsx
tsxCopy code
$this->block_instance->set_attributes_filter_query_args(
    array(
        array(
            'filter'     => 'filter_color',
            'query_type' => 'query_type_color',
        ),
        array(
            'filter'     => 'filter_size',
            'query_type' => 'query_type_size',
        ),
    )
);

```

Therefore, we need to invoke `build_frontend_query` on `$this->block_instance` as follows:

```tsx
$this->block_instance->build_frontend_query( $query, $block, 1 )
```

By doing this, we will have the correct value for `$attributes_filter_query_args`, and our test will pass.

Hence, I believe that none of our tests are giving false positives.

Does that make sense to you as well?

Thanks again. I really appreciate your help 🙏